### PR TITLE
nvwave: Don't close the audio device unless there hasn't been any audio played for 10 seconds.

### LINKED ISSUE #5172: make the audio device still active for better response
**Reported by ondrosik on 2015-06-20 21:39**
Some audio devices have power management system which disables the device when it doesn't play any audio. This makes NVDA less responsive. When you press a key and NVDA wants to say something, you need to wait few miliseconds because the device needs to initialize, and after that play the sound, in this case speech synthesys. This also applies when you for example spell the statusbar or titlebar - spaces between characters are longer. The solution is really simple, Play silence. to the device all the time or somehow make it still active. When you run audio player such as winamp, NVDA is really more responsive, also when you pause the player. I think that we should integrate to NVDA something which will make the soundcard active all the time when NVDA is running. In most situations user can not disable this power management. In older soundcards, there was a comfig file or specific driver related dialog where this was possible, but nowadays it seems that mostly user can't do nothing about it.

### LINKED ISSUE #10721: OneCore performance
### Steps to reproduce:
1. Switch to OneCore.
1. Press windows+m to go to the desktop.
1. Tab to the start button.
1. Shift+tab and tab a few times.

### Actual behavior:
The delay between Desktop list and the icon name varies between fast and slow.
### Expected behavior:
The delay should stay the same.
#### NVDA installed/portable/running from source:
Portable
#### NVDA version:
2019.3 Beta 2
#### Windows version:
Windows 10 build 1909
#### Does the issue still occur after restarting your PC?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
2019.2 works fine.

### PR DESCRIPTION
**Pull request by jcsteh**
### Link to issue number:
Fixes #5172. Fixes #10721 completely.

### Summary of the issue:
1. As described in #5172, some audio drivers/hardware take a long time to open the device and/or truncate the start/end of the audio.
2. As described in https://github.com/nvaccess/nvda/issues/10721#issuecomment-615176345, calling WaveOutOpen in the OneCore synth callback mysteriously blocks (and thus lags) for ~100 ms. Because we close the audio device on idle, we can trigger this problem. Although #11023 mostly fixes this, it's impossible (or at least very difficult) to resolve this completely from within the OneCore driver.
3. Aside from all of this, closing and opening the audio device for rapid short utterances (e.g. rapid movement with the cursor keys or typing) doesn't seem particularly optimal. It's difficult to measure this, but I'd say mitigating this is likely to make audio performance faster/smoother.

### Description of how this pull request fixes the issue:
When WavePlayer.idle() is called and closeWhenIdle is set, instead of closing the device immediately after the audio finishes, set a timer. If audio is played before the timer elapses, stop the timer. Close the device when the timer expires.

### Testing performed:
Tested as per the STR in https://github.com/nvaccess/nvda/issues/10721#issue-554953479.
It's impossible to know when the device is closed, but I placed a beep in WavePlayer._close, had NVDA speak an utterance, then waited 10 seconds and observed that the beep played. I observed that the beep didn't play unless there was a 10 second period of silence.
Anecdotally, response feels a bit snappier with eSpeak when moving rapidly, but this could be placebo effect; I haven't measured this.

### Known issues with pull request:
None.

### CODE REVIEW

**Comment by Brian1Gaff** [comment]
There do seem to be a lot of machines on sale that exhibit this loss of the 
first bit of speech, You almost need a kind of time warp system so the bit 
that you know will get lost is played after the  sound device has struggled 
to its feet, then a little bit of raciness as it catches up.
 I 'd have thought Microsoft ought to get the vendors to allow always on 
sound for desktop machines by default.
 Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal E-mail to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
Newsgroup monitored: alt.comp.blind-users

File: source/nvwave.py
Comment (leonardder) [inline source/nvwave.py:173]: Don't you somehow need to set self._closeTimer to None hereafter? May be I'm just missing something.

**Comment by jcsteh** [comment]
We're going to use the timer again with the same callback, so the only thing setting it to None does is mean we have to recreate it when we next need it, which seems wasteful. Reusing a timer is perfectly valid. Unless I'm missing a concern you have?

**Comment by leonardder** [comment]
> We're going to use the timer again with the same callback, so the only thing setting it to None does is mean we have to recreate it when we next need it, which seems wasteful. Reusing a timer is perfectly valid. Unless I'm missing a concern you have?

Thanks for clarifying.

File: source/nvwave.py
Comment (JulienCochuyt) [inline source/nvwave.py:166]: Wouldn't this function be a candidate for inclusion in `core.py` beside callLater` you added there as of a59e007b35?

File: source/nvwave.py
Comment (JulienCochuyt) [inline source/nvwave.py:305]: Why not use here a regular `threading.Timer` or even `wx.Timer`?
Note `wx.PyTimer` seems to have gone away of wxPython documentation as of the Phoenix migration for some reason (maybe a bad one, no clue)
Sorry if I'm missing the point here.

Comment (jcsteh) [reply source/nvwave.py:305]: threading.Timer would spawn another thread and execute the function there. That's pretty wasteful, since we already have an event loop in our main thread. It also increases the potential for race conditions.
wx.Timer doesn't allow you to pass in a callback, so you have to subclass it. That makes things a lot less readable.

File: source/nvwave.py
Comment (jcsteh) [reply source/nvwave.py:166]: This doesn't notify the watchdog, etc., so it's not safe for general usage. Normally, you'd use queueHandler.queueFunction to run something in the main thread, but I didn't want to deal with the overhead of the core pump, etc. for this case; it really should run ASAP.

**Comment by Neurrone** [comment]
Any chance the default of 10 seconds could be user configurable? For example, there are situations that I might want to set a longer timer and make that trade-off with battery power.

**Comment by k-kolev1985** [comment]
Will this functionality have an option to enable/disable it? Also, as mentioned above, a configurable timeout would also be welcome. Thanks in advance!

**Comment by Adriani90** [comment]
There is no need for optional setting here, this happens in the background and is standard for any screen reader, even voice over plays silence for about 6 seconds but the user does not notice it.
I understand that people fear that this would have significant impact on batery, but this is not the case here. The impact is neglegible.

**Comment by Adriani90** [comment]
A timing interval is also not needed, this would only overload the gui with additional settings which are not standard for any screen reader aat all. I understand that every user wants to have full control over every small setting in the screen reader, but this is not good from a design perspective. But in the end it is Jamie's decision and the decision of NV Access to implement it in whichever way.

**Comment by XLTechie** [comment]
@k-kolev1985, what is the use case for disabling this?

@Neurrone, why would you want to change the timing? Where is the advantage in 
making it longer?

Are you perhaps trying to duplicate the function of the Bluetooth Audio add-on 
by extending this?

**Comment by k-kolev1985** [comment]
@XLTechie The ability to disable it is for battery saving in cases where this NVDA functionality is not needed (e.g. on portable computers which do not exhibit audio cut out behavior). But if really the impact of this functionality on the battery life is negligible, maybe adding the toggle for it it is not worth it.

As for the timeout period configurability. Some people may interrupt their work for longer periods of time but they prefer to still not experience initial speech cut outs when they resume their work.

**Comment by codeofdusk** [comment]
Any updates on this PR?

**Comment by codeofdusk** [comment]
Any updates on this? I suspect it'll help a lot with modern machines that [aggressively close the sound card for power savings](https://mosen.org/soundsfrustrating/).

**Comment by dingpengyu** [comment]
cc @michaelDCurran

**Comment by codeofdusk** [comment]
Suggested changelog entry:

== Changes ==
- NVDA now holds the audio device open for a short time after speech utterances, improving performance on some newer Realtek sound cards. (#5172)

**Comment by leonardder** [comment]
@jcsteh Could you please elaborate on how you fixed the known issue and update the initial description accordingly? Thanks in advance!

**Review by LeonarddeR — DISMISSED** [review]

**Comment by leonardder** [comment]
Has this actually been tested with audio ducking? I assume we need a try build for that to work correctly.

**Comment by jcsteh** [comment]
> @jcsteh Could you please elaborate on how you fixed the known issue

See commit 036761996105b2a9890844b9d44aac238d209693. The timer wasn't being killed in some cases where it needed to be. Now, it is.

> and update the initial description accordingly?

Done.

**Comment by jcsteh** [comment]
> Has this actually been tested with audio ducking? I assume we need a try build for that to work correctly.

I was running this in a try build for weeks (with a bunch of other stuff) and didn't notice any problems with audio ducking. However, it's probably worth double checking, as I wasn't explicitly testing for that.

**Review by feerrenrut — DISMISSED** [review]
This looks fine to me. If no one has any concerns I'd like to go ahead with merging.

**Review by michaelDCurran — DISMISSED** [review]

**Review by feerrenrut — APPROVED** [review]

**Comment by Adriani90** [comment]
I get this error now in last alpha from time to time:
```
ERROR - unhandled exception (23:47:05.138) - MainThread (9068):
Traceback (most recent call last):
  File "wx\core.pyc", line 2196, in Notify
  File "nvwave.pyc", line 349, in _close
  File "nvwave.pyc", line 90, in _winmm_errcheck
OSError: [Errno 5] Das angegebene Gerätehandle ist ungültig.
```


It occurs quite often. Especially in Windows Explorer on Windows 10 2004.

I am using a Dell XPS13 with a Realtec sound card.

Any idea on how to fix this?

**Comment by jcsteh** [comment]
See #11482.

**Comment by feerrenrut** [comment]
@jcsteh Given #11482 I'm tempted to revert this PR and aim for 2020.4 instead. That will take some pressure off fixing this issue and resolving other queries that are popping up.

**Comment by jcsteh** [comment]
Note that #11482 has no user impact apart from the logged error, which won't be noticed in release. I'm not sure what other specific issues you're referring to, but anything related to audio glitches during speech already existed and is unrelated to this PR. That said, this is obviously your call to make.

**Comment by feerrenrut** [comment]
There is also #11490. There have been questions about whether silence should be sent to devices, questions about the impact on power usage and user configuration. In addition I'm not sure I have seen any confirmation that this fixes issues for anyone. Given the short time to release these problems aren't likely to be resolved.

**Comment by Mohamed00** [comment]
As the creator of both issues, I will say that even with the problems it has, I still find that it makes NVDA much more snappy on my machine, and things like rapidly moving through list items seem to be much faster.

**Comment by jcsteh** [comment]
For reference, there's also [this tweet](https://twitter.com/nidza07/status/1292856622008676353?s=21) which suggests it greatly helps another user. Still, I totally understand the desire to mitigate risk; just providing another data point.

**Comment by codeofdusk** [comment]
I strongly vote to keep this PR despite the issues, as the new issues it introduces are (at least in my opinion) far less severe than those fixed by the patch.

However, I definitely think that it's worth considering in next release further improvements to this functionality, particularly playing silence instead of just keeping the device open, with something like #11481 to control the duration of silence.

**Comment by codeofdusk** [comment]
I think the changelog entry might need to be revised though, to focus more on speech performance after brief utterances more than truncation of audio (silence would fix the latter issue, holding open the device seems to only fix the former).

**Comment by nidza07** [comment]
Hello. Yes, I am the author of that tweet, and this PR literally fixes all performance issues I had on my machine with NVDA. Before, progress bar beeps would stutter a lot if played at the same time as speech, now this does not happen at all and the beeps are completely smooth, and as another user correctly mentioned the list navigation is a lot snappier than before. After updating to the alpha where this was merged, it felt like a completely new machine. So for me, the impact of this PR is huge, and I feel it is worth considering keeping this merged unless it causes major issues. I'm not sure if just playing silence is the right solution, since I already have a program doing this on my system for bluetooth devices, and it did not provide any performance improvements, unlike this PR. So thanks once again for your work, and hopefully the issues it has will be resolved.

**Comment by feerrenrut** [comment]
We have done some more testing with this PR, and agree it certainly improves performance significantly in the right circumstances. We will include it in the 2020.3 release, with the intention to fix or mitigate the related issues encountered. Thanks everybody for the extra information provided here.

**Comment by mltony** [comment]
I tested this PR in a recent alpha build and I can report that it also fixes the issue with periodic stuttering in Plantronics backbeat headphones. Playing silence didn't fix stuttering.
However, it doesn't seem to fix the issue of the beginning  of each utterance being lost - this one can be fixed by playing silence, e.g. via my Bluetooth Audio add-on. I would like to offer my help porting this add-on into NVDA core. Please let me know if NVAccess would be interested in such kind of PR.

**Comment by Adriani90** [comment]
Maybe a combination of both approaches would be appropriate here?

Von meinem iPhone gesendet

**Comment by ondrosik** [comment]
I will still use Bluetooth audio addon because I mostly got to a following situation. I sit in a radio studio prepared to read some text. I wait for the song outro (the song is playing from another computer). When It's my time, I press down arrow. NVDA will respond about one second later. I understand that make this configurable may introduce issues mentioned abowe. Maybe we should extend this to one minute but again, some users should come with different situations. I set the bluetooth audio to 10 minutes and simply ignore fact that this Lenovo laptop responds slovly after longer idle time.

**Pull request merged**
